### PR TITLE
Events dead letter queue CH table

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

~~WIP because I still want to re-read this over and scrutinize the schema a bit more.~~

~~Also need to create the topic in Heroku Kafka.~~ ✅ 

**Important:** This doesn't implement the queue anywhere, we're still not producing to this topic. This is the setup on the CH side. Keeping PRs small here

## How did you test this code?

Added 2 tests:

- A test to check inserting to the table works fine. This ensures creating the table will work fine
- A test to check that a message sent to the Kafka topic will be consumed correctly by ClickHouse

### CODE REVIEW

File: .github/workflows/ci-backend.yml
Comment (yakkomajuri) [inline .github/workflows/ci-backend.yml:108]: cc @timgl for this change - this was causing tests to fail for CH migrations

**Comment by yakkomajuri** [comment]
This is ready for a look @fuziontech

File: ee/clickhouse/util.py
Comment (tiina303) [inline ee/clickhouse/util.py:66]: maybe this belongs in a test folder?

File: ee/clickhouse/models/test/test_dead_letter_queue.py
Comment (tiina303) [inline ee/clickhouse/models/test/test_dead_letter_queue.py:77]: Python has setUp and TearDown functions for this

File: ee/clickhouse/models/test/test_dead_letter_queue.py
Comment (guidoiaquinti) [inline ee/clickhouse/models/test/test_dead_letter_queue.py:63]: Instead of hardcoding the values from `TEST_DATA`, what about using `TEST_DATA['event']`?

Comment (yakkomajuri) [reply ee/clickhouse/models/test/test_dead_letter_queue.py:63]: the idea here was to just be explicit in the assertion, as we tend to do with most of our tests. Happy to change though, don't feel strongly about this

File: ee/clickhouse/models/test/test_dead_letter_queue.py
Comment (yakkomajuri) [reply ee/clickhouse/models/test/test_dead_letter_queue.py:77]: They were both supposed to run at the beginning of each test actually, so this was a typo.

But indeed should definitely use `setUp`, thanks!

File: ee/clickhouse/sql/dead_letter_queue.py
Comment (guidoiaquinti) [inline ee/clickhouse/sql/dead_letter_queue.py:5]: If we consider Kafka as the _authoritative source of raw events_ I don’t think we should pipe and duplicate this data in ClickHouse as well. 

If the argument is that we want to have a longer retention period we should probably just configure the new Kafka topic to use a higher [retention.bytes](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#topicconfigs_retention.bytes) or [retention.ms](https://docs.confluent.io/platform/current/installation/configuration/topic-configs.html#topicconfigs_retention.ms) value. I see the point of easier analysis in CH, but we can probably work around that by producing better metrics in the ingestion app.

Last additional note: from a data integrity perspective, trying to synchronise a dataset across two different datastore might lead to higher chances of having data out of sync (as well as duplicating storage costs).

Let me know what do you think about it 🙇

File: ee/clickhouse/models/test/test_dead_letter_queue.py
Comment (guidoiaquinti) [reply ee/clickhouse/models/test/test_dead_letter_queue.py:74]: I didn't know we were doing this also in other places. I'm fine with that, I just thought the other way was maybe less copy paste.

File: ee/clickhouse/sql/dead_letter_queue.py
Comment (yakkomajuri) [reply ee/clickhouse/sql/dead_letter_queue.py:5]: I wouldn't feel super comfortable storing this for too long in Kafka + processing data is much easier in CH.

For instance, with the last issue we had, we got everything out of Kafka and then went through quite some pain to sort the data and get it in the right format. So it's not just about metrics per se, but also other operations we might want to run over the entire dataset. 

With Kafka we have to consume it all first and have it somewhere before being able to do anything about it. 

As for costs, ultimately this table really shouldn't store much, although of course that's not something to rely on. With the outages we had recently that hadn't happened for a while (e.g. Heroku Postgres going down) we would have had 4.5M or so rows here to be kept for a few months max.

Nevertheless, you're more knowledgeable than me here, so happy to get pushback. I also think this approach allows us to move fast though while we work on improving metrics, reliability, etc.

This did make me think about storage though, and realistically I think this could probably have a lower TTL. A month maybe. That should be more than enough time to make decisions on dropping/replaying events.

File: ee/clickhouse/sql/dead_letter_queue.py
Comment (yakkomajuri) [inline ee/clickhouse/sql/dead_letter_queue.py:43]: Updated TTL to 4 weeks based on your feedback about storage @guidoiaquinti 

The topic in Heroku Kafka has 3 day retention (we can change that), so we wouldn't have a lot of duplication, but realistically don't need to keep records in CH for that long.

If we run into a problem like DB going down, then we can hit Kafka directly and produce to the ingestion topic again off a consumer for the DLQ as soon as things are fine.

But if we have a deeper problem that requires some analysis and data parsing then we just default to CH and have more time to come up with a plan.

File: ee/clickhouse/models/test/utils/util.py
Comment (guidoiaquinti) [inline ee/clickhouse/models/test/utils/util.py:9]: What do you think about something like:

```
ts_start = time.time()
timeout = 10

while time.time() < ts_start + timeout:
    result = sync_execute(f"SELECT COUNT(1) FROM {table_name}")
    if result[0][0] == target_row_count:
        return
    sleep(0.5)
```
?

File: ee/clickhouse/sql/dead_letter_queue.py
Comment (guidoiaquinti) [reply ee/clickhouse/sql/dead_letter_queue.py:5]: Let's go on with this implementation so. We can always revisit it in the future.

File: ee/clickhouse/sql/dead_letter_queue.py
Comment (guidoiaquinti) [inline ee/clickhouse/sql/dead_letter_queue.py:47]: Do we have some naming convention regarding table names?

File: ee/clickhouse/util.py
Comment (guidoiaquinti) [inline ee/clickhouse/util.py:2]: Is this import needed? I don't see other related code changes.

File: ee/clickhouse/sql/dead_letter_queue.py
Comment (yakkomajuri) [reply ee/clickhouse/sql/dead_letter_queue.py:47]: For the base tables not really. For kafka tables we always prepend `kafka_` and for materialized views we append `_mv`

File: ee/clickhouse/models/test/utils/util.py
Comment (yakkomajuri) [reply ee/clickhouse/models/test/utils/util.py]: Agree this is clearer, updated!

**Comment by yakkomajuri** [comment]
Any further comments here or ✅ ? @tiina303 @guidoiaquinti

**Comment by guidoiaquinti** [comment]
I don't think I have more comments but as I'm new to this codebase I'll appreciate another pair of 👀 before merging.
Do you mind to add as well to the PR description an e2e test run of this code to verify everything works as expected?

**Comment by yakkomajuri** [comment]
What would constitute e2e in this case? There's no producer for this queue on this PR. This just adds the scaffolding.

Implementing this queue is done in #6230

**Comment by guidoiaquinti** [comment]
> What would constitute e2e in this case? There's no producer for this queue on this PR. This just adds the scaffolding.
> 
> Implementing this queue is done in #6230

Sorry, too much context switching and I've mixed your PRs 🤦 I think this is good to go while for the other we should probably add an e2e test.

**Review by tiina303 — APPROVED** [review]
lgtm (not sure if the ci-backend change)

**Pull request merged**
